v3.3.5: workflow-controller panic at context.go:125 GetTemplateFromRef

<!-- If this is a regression, please open a regression report instead. -->

## Checklist

<!-- Do NOT open an issue until you have: --> 

* [x] Double-checked my configuration.
* [x] Tested using the latest version.
* [x] Used the Emissary executor.

## Summary

Workflow-controller panics suddenly, and once the panic occurs, it continues.
I found other user had same panic in slack.
https://cloud-native.slack.com/archives/C01QW9QSSSK/p1652457569109429

What version are you running?
v3.3.5
This panic was not happend at v3.1.6

## Diagnostics
Note that 
- ClusterWorkflowTemplate certainly exists although it got an error.
- Controller restarts does not fix it.
```
time="2022-06-13T06:04:09.208Z" level=info msg="Get clusterworkflowtemplates 200"
time="2022-06-13T06:04:09.251Z" level=info msg="Processing workflow" namespace=ns1 workflow=wf-lx4zf
time="2022-06-13T06:04:09.251Z" level=info msg="Queueing Failed workflow ns2/wf-kzrs5 for delete in -29s due to TTL"
time="2022-06-13T06:04:09.251Z" level=info msg="Deleting garbage collected workflow 'ns2/wf-kzrs5'"
time="2022-06-13T06:04:09.258Z" level=info msg="Task-result reconciliation" namespace=ns1 numObjs=1 workflow=wf-lx4zf
time="2022-06-13T06:04:09.258Z" level=info msg="task-result changed" namespace=ns1 nodeID=wf-lx4zf-671665593 workflow=wf-lx4zf
time="2022-06-13T06:04:09.259Z" level=error msg="Mark error node" error="clusterworkflowtemplates.argoproj.io \"cluster-template-wf\" not found" namespace=ns1 nodeName="wf-lx4zf[0].wf[0].wf-main-pvc(0)" workflow=wf-lx4zf
time="2022-06-13T06:04:09.259Z" level=info msg="node wf-lx4zf-671665593 phase Running -> Error" namespace=ns1 workflow=wf-lx4zf
time="2022-06-13T06:04:09.259Z" level=info msg="node wf-lx4zf-671665593 message: clusterworkflowtemplates.argoproj.io \"cluster-template-wf\" not found" namespace=ns1 workflow=wf-lx4zf
time="2022-06-13T06:04:09.259Z" level=info msg="node wf-lx4zf-671665593 finished: 2022-06-13 06:04:09.259111158 +0000 UTC" namespace=ns1 workflow=wf-lx4zf
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x19937b6]

goroutine 292 [running]:
github.com/argoproj/argo-workflows/v3/workflow/templateresolution.(*Context).GetTemplateFromRef(0x0, 0xc009b2c030)
	/go/src/github.com/argoproj/argo-workflows/workflow/templateresolution/context.go:125 +0x36
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).GetNodeTemplate(0xc007632240, 0xc00a8b4b00)
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2093 +0x105
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).assessNodeStatus(0xc007632240, 0xc004a16e30, 0xc00a8b4b00)
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1162 +0x9e
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).podReconciliation.func1(0xc004a16e30)
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:997 +0x345
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).podReconciliation.func2(0xc000b94fb8)
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1037 +0x92
created by github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).podReconciliation
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1035 +0x21a"
```


---
<!-- Issue Author: Don't delete this message to encourage other users to support your issue! -->
**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

--- a/workflow/controller/controller.go
+++ b/workflow/controller/controller.go
@@ -420,6 +420,14 @@ func (wfc *WorkflowController) createClusterWorkflowTemplateInformer(ctx context
 	if cwftGetAllowed && cwftListAllowed && cwftWatchAllowed {
 		wfc.cwftmplInformer = informer.NewTolerantClusterWorkflowTemplateInformer(wfc.dynamicInterface, clusterWorkflowTemplateResyncPeriod)
 		go wfc.cwftmplInformer.Informer().Run(ctx.Done())
+
+		// since the above call is asynchronous, make sure we populate our cache before we try to use it later
+		if !cache.WaitForCacheSync(
+			ctx.Done(),
+			wfc.cwftmplInformer.Informer().HasSynced,
+		) {
+			log.Fatal("Timed out waiting for ClusterWorkflowTemplate cache to sync")
+		}
 	} else {
 		log.Warnf("Controller doesn't have RBAC access for ClusterWorkflowTemplates")
 	}

--- a/workflow/controller/controller.go
+++ b/workflow/controller/controller.go
@@ -263,6 +263,7 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, workflowTTLWo
 	go wfc.configMapInformer.Run(ctx.Done())
 	go wfc.wfTaskSetInformer.Informer().Run(ctx.Done())
 	go wfc.taskResultInformer.Run(ctx.Done())
+	wfc.createClusterWorkflowTemplateInformer(ctx)
 
 	// Wait for all involved caches to be synced, before processing items from the queue is started
 	if !cache.WaitForCacheSync(
@@ -276,8 +277,6 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, workflowTTLWo
 	) {
 		log.Fatal("Timed out waiting for caches to sync")
 	}
-
-	wfc.createClusterWorkflowTemplateInformer(ctx)
 
 	// Start the metrics server
 	go wfc.metrics.RunServer(ctx)

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -2100,7 +2100,6 @@ func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) (*wfv1.Templat
 		}
 		return tmpl, nil
 	}
-
 	return woc.wf.GetTemplateByName(node.TemplateName), nil
 }
 

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -1156,10 +1156,6 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 		woc.log.Error(err)
 		return nil
 	}
-	if tmpl == nil {
-		woc.log.Errorf("failed to get old node template corresponding to node status: %v, yet err=nil???", old)
-		return nil
-	}
 	switch pod.Status.Phase {
 	case apiv1.PodPending:
 		new.Phase = wfv1.NodePending

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -1156,6 +1156,10 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 		woc.log.Error(err)
 		return nil
 	}
+	if tmpl == nil {
+		woc.log.Errorf("failed to get old node template corresponding to node status: %v, yet err=nil???", old)
+		return nil
+	}
 	switch pod.Status.Phase {
 	case apiv1.PodPending:
 		new.Phase = wfv1.NodePending

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -1166,7 +1166,7 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 		new.Daemoned = nil
 	case apiv1.PodFailed:
 		// ignore pod failure for daemoned steps
-		if tmpl.IsDaemon() {
+		if tmpl != nil && tmpl.IsDaemon() {
 			new.Phase = wfv1.NodeSucceeded
 		} else {
 			new.Phase, new.Message = woc.inferFailedReason(pod, tmpl)
@@ -1176,7 +1176,7 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 		// Daemons are a special case we need to understand the rules:
 		// A node transitions into "daemoned" only if it's a daemon template and it becomes running AND ready.
 		// A node will be unmarked "daemoned" when its boundary node completes, anywhere killDaemonedChildren is called.
-		if tmpl.IsDaemon() {
+		if tmpl != nil && tmpl.IsDaemon() {
 			if !old.Fulfilled() {
 				// pod is running and template is marked daemon. check if everything is ready
 				for _, ctrStatus := range pod.Status.ContainerStatuses {

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -2099,7 +2099,6 @@ func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) *wfv1.Template
 		}
 		return tmpl
 	}
-
 	return woc.wf.GetTemplateByName(node.TemplateName)
 }
 

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -1151,8 +1151,8 @@ func printPodSpecLog(pod *apiv1.Pod, wfName string) {
 // and returns the new node status if something changed
 func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus) *wfv1.NodeStatus {
 	new := old.DeepCopy()
-	tmpl := woc.GetNodeTemplate(old)
-	if tmpl == nil {
+	tmpl, err := woc.GetNodeTemplate(old)
+	if err != nil {
 		return nil
 	}
 	switch pod.Status.Phase {
@@ -2086,20 +2086,21 @@ func (woc *wfOperationCtx) hasDaemonNodes() bool {
 	return false
 }
 
-func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) *wfv1.Template {
+func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) (*wfv1.Template, error) {
 	if node.TemplateRef != nil {
 		tmplCtx, err := woc.createTemplateContext(node.GetTemplateScope())
 		if err != nil {
 			woc.markNodeError(node.Name, err)
-			return nil
+			return nil, err
 		}
 		tmpl, err := tmplCtx.GetTemplateFromRef(node.TemplateRef)
 		if err != nil {
 			woc.markNodeError(node.Name, err)
+			return tmpl, err
 		}
-		return tmpl
+		return tmpl, nil
 	}
-	return woc.wf.GetTemplateByName(node.TemplateName)
+	return woc.wf.GetTemplateByName(node.TemplateName), nil
 }
 
 func (woc *wfOperationCtx) markWorkflowRunning(ctx context.Context) {

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -2100,6 +2100,7 @@ func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) (*wfv1.Templat
 		}
 		return tmpl, nil
 	}
+
 	return woc.wf.GetTemplateByName(node.TemplateName), nil
 }
 

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -1152,6 +1152,9 @@ func printPodSpecLog(pod *apiv1.Pod, wfName string) {
 func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus) *wfv1.NodeStatus {
 	new := old.DeepCopy()
 	tmpl := woc.GetNodeTemplate(old)
+	if tmpl == nil {
+		return nil
+	}
 	switch pod.Status.Phase {
 	case apiv1.PodPending:
 		new.Phase = wfv1.NodePending
@@ -2088,6 +2091,7 @@ func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) *wfv1.Template
 		tmplCtx, err := woc.createTemplateContext(node.GetTemplateScope())
 		if err != nil {
 			woc.markNodeError(node.Name, err)
+			return nil
 		}
 		tmpl, err := tmplCtx.GetTemplateFromRef(node.TemplateRef)
 		if err != nil {

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -2099,6 +2099,7 @@ func (woc *wfOperationCtx) GetNodeTemplate(node *wfv1.NodeStatus) *wfv1.Template
 		}
 		return tmpl
 	}
+
 	return woc.wf.GetTemplateByName(node.TemplateName)
 }
 

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -1153,6 +1153,7 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, old *wfv1.NodeStatus
 	new := old.DeepCopy()
 	tmpl, err := woc.GetNodeTemplate(old)
 	if err != nil {
+		woc.log.Error(err)
 		return nil
 	}
 	switch pod.Status.Phase {

